src/security/Handshake.cc

```diff
@@ -274,24 +274,30 @@ Security::HandshakeParser::parseModernRecord()
     Must(record.fragment.length() || record.type == ContentType::ctApplicationData);
 
     if (currentContentType != record.type) {
+        tkMessages.expectMore(false);
         parseMessages();
         Must(tkMessages.atEnd()); // no currentContentType leftovers
-        fragments = record.fragment;
         currentContentType = record.type;
-    } else {
-        fragments.append(record.fragment);
     }
 
-    if (tkRecords.atEnd() && !done)
-        parseMessages();
+    const auto haveUnparsedRecordBytes = !tkRecords.atEnd();
+    const auto expectMoreRecordLayerBytes = tkRecords.expectMore();
+    // TODO: consider adding BinaryTokenizer::exhausted() instead
+    const auto expectMoreMessageLayerBytes = haveUnparsedRecordBytes || expectMoreRecordLayerBytes;
+
+    tkMessages.expectMore(expectMoreMessageLayerBytes);
+    tkMessages.append(record.fragment);
+
+    parseMessages();
 }
 
 /// parses one or more "higher-level protocol" frames of currentContentType
 void
 Security::HandshakeParser::parseMessages()
 {
-    tkMessages.reset(fragments, false);
-    for (; !tkMessages.atEnd(); tkMessages.commit()) {
+    tkMessages.rollback();
+
+    while (!tkMessages.atEnd() && !done) {
```

Comment:
We probably need to ensure somehow that tkMessages.commit() was called after each iteration.

Reply:
> agree that Option 3 is the best way forward

Yes, I think it is the best among other options that you outlined. Done (32c24e5).